pkg/snapshot/restorer/restorer.go

```diff
@@ -317,24 +326,260 @@ func startEmbeddedEtcd(ro RestoreOptions) (*embed.Etcd, error) {
+func (r *Restorer) applyDeltaSnapshots(client *clientv3.Client, ro RestoreOptions) error {
+	snapList := ro.DeltaSnapList
+	numMaxFetchers := ro.MaxFetchers
+
+	// no delta snapshots available
+	if len(snapList) == 0 {
+		return nil
+	}
+
 	firstDeltaSnap := snapList[0]
+
+	r.logger.Infof("Applying delta snapshot %s", path.Join(firstDeltaSnap.SnapDir, firstDeltaSnap.SnapName))
 	if err := r.applyFirstDeltaSnapshot(client, *firstDeltaSnap); err != nil {
 		return err
 	}
 	if err := verifySnapshotRevision(client, snapList[0]); err != nil {
 		return err
 	}
-	for _, snap := range snapList[1:] {
-		if err := r.applyDeltaSnapshot(client, *snap); err != nil {
-			return err
+
+	// no more delta snapshots available
+	if len(snapList) == 1 {
+		return nil
+	}
+
+	var (
+		remainingSnaps         = snapList[1:]
+		numSnaps               = len(remainingSnaps)
+		snapLocations          = make([]string, numSnaps)
+		errCh                  = make(chan error, 1)
+		snapInfoCh             = make(chan snapInfo, numSnaps)
+		fetchedEventsInfoCh    = make(chan eventsInfo, numSnaps)
+		applyInfoCh            = make(chan applyInfo, numSnaps)
+		numFetchers            = int(math.Min(float64(numMaxFetchers), float64(numSnaps)))
+		stopCh                 = make(chan bool, numFetchers+2) // used for stopping all routines (fetchers, persister and applier)
+		fetchersStoppedCh      = make(chan bool, numFetchers)   // used to indicate that fetchers have stopped
+		persistDeltasStoppedCh = make(chan bool, numFetchers)   // used to indicate that persister has stopped
+		applyStoppedCh         = make(chan bool, numFetchers)   // used to indicate that applier has stopped
+	)
+
+	go r.persistDeltaSnapshots(numSnaps, snapLocations, fetchedEventsInfoCh, applyInfoCh, errCh, stopCh, persistDeltasStoppedCh)
+	go r.apply(client, remainingSnaps, applyInfoCh, errCh, stopCh, applyStoppedCh)
+
+	for f := 0; f < numFetchers; f++ {
+		go r.fetch(f, snapInfoCh, fetchedEventsInfoCh, errCh, stopCh, fetchersStoppedCh)
+	}
+
+	for i, snap := range remainingSnaps {
+		snapInfo := snapInfo{
+			Snapshot:  *snap,
+			SnapIndex: i,
 		}
-		if err := verifySnapshotRevision(client, snap); err != nil {
+		snapInfoCh <- snapInfo
+	}
+	close(snapInfoCh)
+
+	for {
+		select {
+		case err := <-errCh:
+			r.cleanup(numFetchers, snapLocations, stopCh, fetchersStoppedCh, persistDeltasStoppedCh, applyStoppedCh)
+			if err == nil {
+				r.logger.Infof("Restoration complete.")
+				return nil
+			}
+			r.logger.Warnf("Restoration failed.")
 			return err
+		default:
+			time.Sleep(time.Millisecond * 10)
+		}
+	}
+}
+
+// cleanup stops all running goroutines and removes the persisted snapshot files from disk
+func (r *Restorer) cleanup(numFetchers int, snapLocations []string, stopCh chan<- bool, fetchersStoppedCh <-chan bool, persistDeltasStoppedCh <-chan bool, applyStoppedCh <-chan bool) {
+	for i := 0; i < numFetchers+2; i++ {
```

Comment:
Use `waitGroup`. 
FYI,  for such case where you wan't send stop signal to everybody who are sharing same channel and the value in channel doesn't matter for them, just use `close(stopCh)`. No need of buffered `stopCh` and for loop sending individual stop channel to each fetcher.

Reply:
Sure. Thanks! Have implemented this now.  :)